File: dashboard/src/lib/types/contracts.ts
// Single source of truth for cross-boundary contracts (D1 §6).
// Mirrors the Rust types in src-tauri/src/contracts.rs — keep them in sync.

export type StepId = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;

export type Severity = "CRITICAL" | "MAJOR" | "MINOR" | "ADVISORY";
export type GateDecision = "PASS" | "FAIL" | "ARBITRATION";
export type LoopMode = "full-auto" | "semi-auto" | "supervised";

export interface Finding {
  id: string;
  severity: Severity;
  what: string;
  why: string;
  how_to_fix: string;
  status: "OPEN" | "RESOLVED" | "ACCEPTED_RISK";
}

export interface CriticScore {
  critic_id: string;
  score: number;
  valid: boolean;
  invalidation_reason: string | null;
  key_condition: string;
  findings: Finding[];
}

// The gate's only input (R7).
export interface ScoreReport {
  phase: number;
  iteration: number;
  phase_score: number;
  decision: GateDecision;
  bottleneck_critic: string | null;
  next_action: string;
  critic_scores: CriticScore[];
}

export type StructuralMarker =
  | { kind: "builderPersona"; name: string }
  | { kind: "critiqueSection"; critic: string }
  | { kind: "scoreCardStart" }
  | { kind: "remediationTicket"; id: string }
  | { kind: "phaseExit" };

// The single backend -> frontend seam (ADR-005).
export type AgentEvent =
  | { type: "stepStart"; runId: string; phaseId: string; stepId: StepId; iteration: number }
  | { type: "token"; runId: string; stepId: StepId; chunk: string; seq: number }
  | { type: "marker"; runId: string; stepId: StepId; marker: StructuralMarker }
  | { type: "stepComplete"; runId: string; stepId: StepId; artifactRef: string }
  | { type: "score"; runId: string; phaseId: string; report: ScoreReport }
  | { type: "fileWritten"; runId: string; path: string; phaseId: string }
  | { type: "phasePassed"; runId: string; phaseId: string; snapshotRef: string }
  | { type: "needsHuman"; runId: string; reason: string; payload: unknown }
  | { type: "error"; runId: string; stepId: StepId | null; error: string; recoverable: boolean };

export type Platform = "web" | "mobile" | "api" | "desktop" | "embedded";
export type Scale = "prototype" | "startup" | "enterprise";
export type Compliance = "none" | "GDPR" | "HIPAA" | "SOC2" | "PCI";
export type Preset = "Lite" | "Standard" | "Full";

export interface Phase0Seed {
  idea: string;
  context: {
    platform: Platform[];
    scale: Scale;
    compliance: Compliance[];
    constraints?: string;
  };
  preset: Preset;
  // Phase A express path: a single model for every step. Keys never travel here.
  model: { providerId: string; model: string };
}

export interface ProjectMeta {
  id: string;
  name: string;
  preset: Preset;
  currentPhase: string;
  lastActive: string;
}

export type RouteRole = "continuity" | "research" | "builder" | "critique" | "scoring" | "arbitration" | "snapshot";
export type ContextBand = "Healthy" | "Amber" | "Red" | "Critical";

export interface ModelCapabilities {
  jsonMode: boolean;
  functionCalling: boolean;
  streaming: boolean;
  estContextWindow: number;
}

export interface ProviderRef {
  id: string;
  kind: string;
  label: string;
  baseUrl: string;
  model: string;
  capabilities: ModelCapabilities;
}

export interface ContextBudget {
  usedTokens: number;
  limitTokens: number;
  ratio: number;
  band: ContextBand;
  recommendation: string;
}

export interface RoutePlan {
  role: RouteRole;
  providerId: string;
  providerKind: string;
  model: string;
  reason: string;
  requirements: string[];
  capabilities: ModelCapabilities;
  budget: ContextBudget;
}

export interface RoutingOverview {
  preset: Preset;
  routes: RoutePlan[];
}

export type AuditVerdict = "Pass" | "Warn";

export interface AuditFinding {
  label: string;
  verdict: AuditVerdict;
  detail: string;
}

export interface AuditSection {
  name: string;
  findings: AuditFinding[];
}

export interface DashboardAuditReport {
  overall: AuditVerdict;
  security: AuditSection;
  performance: AuditSection;
  accessibility: AuditSection;
}

File: dashboard/core/src/contracts.rs
//! Rust mirror of `src/lib/types/contracts.ts`. Keep the two in sync.

use serde::{Deserialize, Serialize};

pub type StepId = u8;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum Severity {
    Critical,
    Major,
    Minor,
    Advisory,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum GateDecision {
    Pass,
    Fail,
    Arbitration,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum FindingStatus {
    Open,
    Resolved,
    AcceptedRisk,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Finding {
    pub id: String,
    pub severity: Severity,
    pub what: String,
    pub why: String,
    pub how_to_fix: String,
    pub status: FindingStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CriticScore {
    pub critic_id: String,
    pub score: f64,
    pub valid: bool,
    pub invalidation_reason: Option<String>,
    pub key_condition: String,
    #[serde(default)]
    pub findings: Vec<Finding>,
}

/// The gate's only input (R7).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScoreReport {
    pub phase: i64,
    pub iteration: i64,
    pub phase_score: f64,
    pub decision: GateDecision,
    pub bottleneck_critic: Option<String>,
    pub next_action: String,
    pub critic_scores: Vec<CriticScore>,
}

impl ScoreReport {
    /// Phase score = lowest VALID critic score (no averaging). Unsupported scores
    /// (no findings backing a sub-9 score is allowed; a >=9 score with open CRITICAL
    /// findings is invalidated by the aggregator). This recomputes from valid scores.
    pub fn lowest_valid_score(&self) -> Option<f64> {
        let mut lowest: Option<f64> = None;
        for score in self.critic_scores.iter().filter(|c| c.valid).map(|c| c.score) {
            lowest = Some(lowest.map_or(score, |current| current.min(score)));
        }
        lowest
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum StructuralMarker {
    BuilderPersona { name: String },
    CritiqueSection { critic: String },
    ScoreCardStart,
    RemediationTicket { id: String },
    PhaseExit,
}

/// The single backend -> frontend seam (ADR-005).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum AgentEvent {
    StepStart { run_id: String, phase_id: String, step_id: StepId, iteration: i64 },
    Token { run_id: String, step_id: StepId, chunk: String, seq: u64 },
    Marker { run_id: String, step_id: StepId, marker: StructuralMarker },
    StepComplete { run_id: String, step_id: StepId, artifact_ref: String },
    Score { run_id: String, phase_id: String, report: ScoreReport },
    FileWritten { run_id: String, path: String, phase_id: String },
    PhasePassed { run_id: String, phase_id: String, snapshot_ref: String },
    NeedsHuman { run_id: String, reason: String, payload: serde_json::Value },
    Error { run_id: String, step_id: Option<StepId>, error: String, recoverable: bool },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Phase0Context {
    pub platform: Vec<String>,
    pub scale: String,
    pub compliance: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub constraints: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelSelection {
    pub provider_id: String,
    pub model: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Phase0Seed {
    pub idea: String,
    pub context: Phase0Context,
    pub preset: String,
    /// Express path: one model for every step. Keys never travel in this struct.
    pub model: ModelSelection,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectMeta {
    pub id: String,
    pub name: String,
    pub preset: String,
    pub current_phase: String,
    pub last_active: String,
}
